src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java

```diff
@@ -109,19 +110,48 @@ protected void deinitializeModule() throws Exception {
 
         Map<String, TokenType> tokens = new HashMap<>();
 
-        for (int slotIndex = 0; slotIndex < slots.length; slotIndex++) {
-            TokenType token = createToken(slots, slotIndex);
-            TokenType previous = tokens.putIfAbsent(token.getId(), token);
-
-            if (previous == null) {
-                log.info("Module '{}' slot #{} has token with ID '{}': {}", module.getType(), slotIndex, token.getId(),
-                        token);
-            } else {
-                log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
-                        module.getType(), slotIndex, token.getId());
+        // HSM slots defined in signer configuration
+        String[] hsmSlots = SystemProperties.getHSMSlotIndexes();
+        // Scan all available HSM slots
+        if (hsmSlots.length == 0) {
+            for (int slotIndex = 0; slotIndex < slots.length; slotIndex++) {
+                TokenType token = createToken(slots, slotIndex);
+                TokenType previous = tokens.putIfAbsent(token.getId(), token);
+
+                if (previous == null) {
+                    log.info("Module '{}' slot #{} has token with ID '{}': {}", module.getType(), slotIndex,
+                    token.getId(), token);
+                } else {
+                    log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
+                            module.getType(), slotIndex, token.getId());
+                }
+            }
+        // Only scan defined slots if HSM_SLOT_INDEXES variable configured
+        } else {
+            for (int i = 0; i < hsmSlots.length; i++) {
+                if (hsmSlots[i].matches("\\d+")) {
```

Comment:
Some invalid values are ignored and skipped, where others (too large for an int) will throw an exception and stop the processing. This is inconsistent, the check should either skip all invalid values or an exception should be thrown.